volume/local/local.go

```diff
@@ -79,13 +79,17 @@ func New(scope string, rootIdentity idtools.Identity) (*Root, error) {
 			quotaCtl:   r.quotaCtl,
 		}
 
-		// unmount anything that may still be mounted (for example, from an
-		// unclean shutdown). This is a no-op on windows
-		unmount(v.path)
-
 		if err := v.loadOpts(); err != nil {
 			return nil, err
 		}
+
+		// Unmount data directory if might been mounted by the daemon and
+		// possibly wasn't unmounted (for example due to an unclean shutdown).
+		// This is a no-op on Windows.
+		if v.needsMount() {
+			unmount(v.path)
```

Comment:
Oh, hmmm...
We just fixed a bug (or I guess partially?) with live-restore and mount refcounting where we just assume the volume is still mounted.
Not likely a problem for many (especially since no one has said anything about it) but, maybe we shouldn't be unmounting here and instead always restore mount ref counts?

When containers are shutdown as part of a normal startutp these would get unmounted anyway.

Reply:
Not sure I follow. If we'd just "accept" that this is already mounted and just increment a refcount, `Unmount` won't unmount and decrement the active count anyway:
https://github.com/moby/moby/blob/a41f9c38b314da99cb0eb63176a54baef70437df/volume/local/local.go#L328-L336

Reply:
Right but now we assume it's mounted and won't mount it again.

Reply:
Oh, right, I tested it and the live restore of the volumes with mount opts is broken - the `_data` directory is unmounted here, so any new container which mounts that volume mounts the empty `_data` directory instead.

Looks like we shouldn't do unmount, at least not when the live restore is enabled. Although I don't think we can easily check if live-restore is enabled in the `local.New`.

Reply:
Opened: https://github.com/moby/moby/pull/46253
Leaving this one to not lose the context.